### PR TITLE
Remove document options provider

### PR DESCRIPTION
**Pull request by tmat**
Builds on top of recent options changes and removes `IDocumentOptionsProvider`. This is internal extensibility point that allowed define custom options for documents. It was only used for Razor and inferred indentation, both of which are now addressed differently.

### CODE REVIEW

File: src/Workspaces/CoreTest/Formatter/FormatterTests.cs [hunk 1]
```diff
@@ -101,16 +64,10 @@ public async Task FormatAsync_ForeignLanguageWithFormattingSupport()
     [CombinatorialData]
     public async Task FormatAsync_ForeignLanguageWithFormattingSupport_Options(bool passExplicitOptions)
```
Comment (sharwell) [inline src/Workspaces/CoreTest/Formatter/FormatterTests.cs:65]: ❔ Is this test still providing value?

File: src/EditorFeatures/CSharpTest/Formatting/RazorLineFormattingOptionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            => typeof(TService) == typeof(DocumentPropertiesService) ? (TService?)(object)new PropertiesService() : null;
+
+        internal sealed class PropertiesService : DocumentPropertiesService
+        {
+            public override string? DiagnosticsLspClientName => "RazorCSharp";
+        }
+    }
+
+    [Fact]
+    [UseExportProvider]
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Formatting/RazorLineFormattingOptionsTests.cs:45]: 💡  This attribute is typically applied to the containing type

**Review by dibarbet — APPROVED** [review]

File: src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs [hunk 3]
```diff
@@ -15,28 +20,79 @@ namespace Microsoft.CodeAnalysis.Options
     /// </summary>
     public sealed class DocumentOptionSet : OptionSet
     {
-        private readonly OptionSet _backingOptionSet;
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs:18]: 📝 I didn't fully verify the correctness of inlining this as `DocumentSpecificOptionSet`

File: src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs [hunk 4]
```diff
@@ -15,28 +20,79 @@ namespace Microsoft.CodeAnalysis.Options
...
+        private ImmutableDictionary<OptionKey, object?> _values;
         private readonly string _language;
 
-        internal DocumentOptionSet(OptionSet backingOptionSet, string language)
+        internal DocumentOptionSet(StructuredAnalyzerConfigOptions? configOptions, OptionSet underlyingOptions, string language)
+            : this(configOptions, underlyingOptions, language, ImmutableDictionary<OptionKey, object?>.Empty)
+        {
+        }
+
+        internal DocumentOptionSet(StructuredAnalyzerConfigOptions? configOptions, OptionSet underlyingOptions, string language, ImmutableDictionary<OptionKey, object?> values)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs:33]: ❔ Can this constructor be `private`?

**Review by sharwell — APPROVED** [review]

File: src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Options/DocumentOptionSet.cs:33]: Yup.

**Pull request merged**